### PR TITLE
Add off-by-one variation for Softmax, as well as temperature and input saturation

### PR DESCRIPTION
**Pull request by klei22**
Create learned variations for off-by-one and temperature, adding more options for the strongermax softmax variation template.

Modify the sweep to include single dimension variations for each of the major strongermax options.

### CODE REVIEW

File: explorations/strongermax_sweep.json
Comment (gkielian) [inline explorations/strongermax_sweep.json:1]: Please refactor this to utilize parameter groups.

**Review by gkielian — CHANGES_REQUESTED** [review]
Looks good, hoping we can utilize param groups to shorten total size of the strongermax_sweep.

**Comment by klei22** [comment]
# Results

image

- Learned off by one improves val loss, and even more than obo alone.
- Sharper bases improve validation loss (temperature increases or direct base increases)
- Clamping appears to hurt the val loss, with diminishing effect as we approach 88 (though 50 was slight).

## Learned Off By One Improves Val Loss
Finding that we're able to significantly improve the validation loss via learned-off-by-one.

## Learned Temperature Causes Instability
The learned temperature led to NaN's before any checkpoints (before iteration 250 in this case), but the non-learned versions did converge and some led to better validation losses (seemingly preferring steeper values).

## Clamping
Clamping to 25 50ish and 88 were tried, and the larger the clamp value the better.
We'd like to not have to clamp at 88, since this is already pretty high.
We should try to clamp at values from 88 to 0, and draw a curve of the validation loss after x iterations.

**Review by gkielian — APPROVED** [review]
LGTM
